syntax = "proto3";

import "google/protobuf/empty.proto";

service Users {
    rpc GetUserStatus (UserId) returns (OptionalUserStatus) {}
    rpc SetUserStatus (SetStatusRequest) returns (Response) {}
    rpc TransferUsers (UsersMap) returns (Response) {}
    rpc RequestBackup (google.protobuf.Empty) returns (UsersMap) {}
}

message UserId {
    int32 user_id = 1;
}

message OptionalUserStatus {
    Status status = 1;
    bool exists = 2;
}

message SetStatusRequest {
    int32 user_id = 1;
    Status status = 2;
}

message Response {
    bool success = 1;
}

message UsersMap {
    map<int32, Status> users = 1;
}

enum Status {
    ONLINE = 0;
    OFFLINE = 1;
}